[FreeType] Centralize FcPattern creation

donny-dont opened the pull request:
#### 76d359b2b1931a8e41493b236af40729ba254fbe
<pre>
[FreeType] Centralize FcPattern creation
<a href="https://bugs.webkit.org/show_bug.cgi?id=267610">https://bugs.webkit.org/show_bug.cgi?id=267610</a>

Reviewed by NOBODY (OOPS!).

Explanation of why this fixes the bug (OOPS!).

* Source/WebCore/platform/graphics/freetype/FontCacheFreeType.cpp:
* Source/WebCore/platform/graphics/freetype/FontCustomPlatformDataFreeType.cpp:
* Source/WebCore/platform/graphics/freetype/FontPlatformDataFreeType.cpp:
* Source/WebCore/platform/graphics/freetype/FontSetCache.cpp:
* Source/WebCore/platform/graphics/freetype/FontconfigUtilities.cpp: Added.
* Source/WebCore/platform/graphics/freetype/FontconfigUtilities.h: Added.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/76d359b2b1931a8e41493b236af40729ba254fbe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34322 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13128 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36313 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37000 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31061 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35402 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15504 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10242 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30072 "Failure limit exceed. At least found 496 new test failures: accessibility/checkbox-radio-element-rect.html, accessibility/content-editable-as-textarea.html, accessibility/dynamic-changes-update-position.html, accessibility/image-link.html, accessibility/image-map2.html, accessibility/lists.html, accessibility/node-only-object-element-rect.html, accessibility/press-targets-center-point.html, accessibility/table-attributes.html, accessibility/table-cell-spans.html ... (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34834 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11105 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30607 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9695 "Found 2 new API test failures: /TestWebCore:ComplexTextControllerTest.TotalWidthWithJustification, /TestWebCore:ComplexTextControllerTest.InitialAdvanceInRTLNoOrigins (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9788 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30653 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38284 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31168 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30974 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35867 "Failure limit exceed. At least found 498 new test failures: accessibility/checkbox-radio-element-rect.html, accessibility/content-editable-as-textarea.html, accessibility/dynamic-changes-update-position.html, accessibility/gtk/object-attributes.html, accessibility/gtk/text-at-offset-wrapped-lines.html, accessibility/gtk/xml-roles-exposed.html, accessibility/image-link.html, accessibility/image-map2.html, accessibility/lists.html, accessibility/press-targets-center-point.html ... (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9916 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7790 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33812 "Found 2 new API test failures: /TestWebCore:ComplexTextControllerTest.TotalWidthWithJustification, /TestWebCore:ComplexTextControllerTest.InitialAdvanceInRTLNoOrigins (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11712 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10473 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10753 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebCore/platform/graphics/freetype/FontconfigUtilities.cpp:
```diff
@@ -0,0 +1,32 @@
+
+namespace WebCore {
+
+RefPtr<FcPattern> fontconfigPatternMatcher()
```
You forgot to rename this.

mcatanzaro commented on Source/WebCore/platform/graphics/freetype/FontconfigUtilities.cpp:
```diff
@@ -0,0 +1,57 @@
+        cairo_ft_font_options_substitute(getDefaultCairoFontOptions(), pattern);
+#endif
+        FcDefaultSubstitute(pattern);
+        FcPatternDel(pattern, FC_FAMILY);
+        FcConfigSubstitute(nullptr, pattern, FcMatchFont);
```
You really need to preserve the "Get some generic default settings from fontconfig for web fonts" comment because this is gobbledygook and looks like it's surely a mistake without the comment to explain it.

donny-dont committed:
[FreeType] Centralize FcPattern creation
https://bugs.webkit.org/show_bug.cgi?id=267610

Reviewed by NOBODY (OOPS!).

Explanation of why this fixes the bug (OOPS!).

* Source/WebCore/platform/graphics/freetype/FontCacheFreeType.cpp:
* Source/WebCore/platform/graphics/freetype/FontCustomPlatformDataFreeType.cpp:
* Source/WebCore/platform/graphics/freetype/FontPlatformDataFreeType.cpp:
* Source/WebCore/platform/graphics/freetype/FontSetCache.cpp:
* Source/WebCore/platform/graphics/freetype/FontconfigUtilities.cpp: Added.
* Source/WebCore/platform/graphics/freetype/FontconfigUtilities.h: Added.
